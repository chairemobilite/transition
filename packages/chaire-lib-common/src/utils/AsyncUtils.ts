/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

const causeTimeout = 'PROMISE_TIMEOUT';
export const isTimeout = (error: unknown) => error && (error as any).cause && (error as any).cause === causeTimeout;

/**
 * Call an async function with a maximum time limit (in milliseconds) for the timeout
 * Taken from https://javascript.plainenglish.io/how-to-add-a-timeout-limit-to-asynchronous-javascript-functions-3676d89c186d
 * @param {Promise<T>} asyncPromise An asynchronous promise to resolve
 * @param {number} timeLimit Time limit to attempt function in milliseconds
 * @returns {<T> | undefined} Resolved promise for async function call, or an error if time limit reached
 */
export const asyncCallWithTimeout = async <T>(asyncPromise: Promise<T>, timeLimit: number): Promise<T> => {
    let timeoutHandle: NodeJS.Timeout;

    const timeoutPromise = new Promise<T>((_resolve, reject) => {
        timeoutHandle = setTimeout(() => {
            reject(new Error('Async call timeout limit reached', { cause: causeTimeout }));
        }, timeLimit);
    });

    return Promise.race([asyncPromise, timeoutPromise])
        .then((result) => {
            clearTimeout(timeoutHandle);
            return result;
        })
        .catch((error) => {
            if (!isTimeout(error)) {
                // Error in promise, not because of timeout
                clearTimeout(timeoutHandle);
            }
            throw error;
        });
};
