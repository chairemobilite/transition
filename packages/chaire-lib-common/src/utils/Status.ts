/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
/* Collection of status objects to properly type json objects */

export interface StatusError {
    status: 'error';
    error: unknown;
}

export interface StatusResult<T> {
    status: 'ok';
    result: T;
}

export type Status<T> = StatusResult<T> | StatusError;

// Function to easy creation of status objects
export function createError(error: unknown): StatusError {
    return { status: 'error', error: error };
}

export function createOk<T>(result: T): StatusResult<T> {
    return { status: 'ok', result: result };
}

// Type guards
export function isStatusError<T>(status: Status<T>): status is StatusError {
    return status.status === 'error';
}

export function isStatusOk<T>(status: Status<T>): status is StatusResult<T> {
    return status.status === 'ok';
}

// Losely inspired by Rust
export function unwrap<T>(status: Status<T>): T {
    if (isStatusOk(status)) {
        return status.result;
    } else if (isStatusError(status)) {
        throw status.error;
    } else {
        // We did not receive a Status object
        throw 'Invalid Status object';
    }
}
