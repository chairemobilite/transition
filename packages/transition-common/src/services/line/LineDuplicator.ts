/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { Line } from './Line';
import { Path as TransitPath } from '../path/Path';
import { duplicateService } from '../service/ServiceDuplicator';
import { duplicateSchedule } from '../schedules/ScheduleDuplicator';

export interface DuplicateLineOptions {
    socket: any;
    duplicateSchedules?: boolean;
    duplicateServices?: boolean;
    agencyId?: string | false;
    serviceIdsMapping?: { [key: string]: string };
    newShortname?: string;
    newLongname?: string;
    newServiceSuffix?: string;
}

export const duplicateLine = async (
    baseLine: Line,
    {
        socket,
        duplicateSchedules = false,
        duplicateServices = false,
        agencyId = false,
        serviceIdsMapping = {},
        newShortname = '',
        newLongname = '',
        newServiceSuffix = ''
    }: DuplicateLineOptions
) => {
    // TODO tahini: the duplication process should be a transaction. If a database error occurs, it should not proceed.

    const oldNewServiceidsMapping = serviceIdsMapping || {};

    baseLine.refreshPaths();
    const pathsCount = baseLine.paths.length;
    const collectionManager = baseLine.collectionManager;

    // clone line and change or reinitialize the attributes that need to be changed (eg. schedules and paths)
    const newAttributes = baseLine.getClonedAttributes(true);
    if (agencyId) {
        newAttributes.agency_id = agencyId;
    }
    newAttributes.shortname = `${newShortname ? newShortname : newAttributes.shortname}`;
    newAttributes.longname = `${newLongname ? newLongname : newAttributes.longname}`;
    const duplicatedLine = new Line(newAttributes, true, collectionManager);

    // Save the line a first time to have its id in the database
    await duplicatedLine.save(socket);
    const lineCollection = collectionManager.get('lines');
    lineCollection.add(duplicatedLine);

    const oldPaths = baseLine.paths;
    const newPathIds: string[] = [];
    const oldNewPathIdsMapping = {}; // for schedules

    for (let i = 0; i < pathsCount; i++) {
        const oldPath = oldPaths[i];
        const newPathAttributes = oldPath.getClonedAttributes(true);

        newPathAttributes.line_id = duplicatedLine.getId();
        const duplicatedPath = new TransitPath(newPathAttributes, true, collectionManager);
        await duplicatedPath.save(socket);
        newPathIds.push(duplicatedPath.getId());
        oldNewPathIdsMapping[oldPath.getId()] = duplicatedPath.getId();
        collectionManager.get('paths').add(duplicatedPath);
    }

    duplicatedLine.attributes.path_ids = newPathIds;
    duplicatedLine.refreshPaths();

    const serviceIdsUsed: string[] = [];
    const newDuplicatedSchedules = {};

    // duplicate schedules:
    if (duplicateSchedules) {
        await baseLine.refreshSchedules(socket);

        const oldSchedules = baseLine.getSchedules();

        for (const serviceId in oldSchedules) {
            let newServiceId = serviceId;

            if (duplicateServices && oldNewServiceidsMapping[serviceId]) {
                newServiceId = oldNewServiceidsMapping[serviceId];
            } else if (duplicateServices) {
                const oldService = collectionManager.get('services').getById(serviceId);
                const newService = await duplicateService(oldService, {
                    socket,
                    newServiceSuffix,
                    serviceCollection: collectionManager.get('services')
                });
                newServiceId = newService.getId();
                oldNewServiceidsMapping[serviceId] = newServiceId;
            }

            const oldSchedule = oldSchedules[serviceId];
            const newSchedule = await duplicateSchedule(oldSchedule, {
                lineId: duplicatedLine.getId(),
                serviceId: newServiceId,
                pathIdsMapping: oldNewPathIdsMapping
            });
            await newSchedule.save(socket);
            duplicatedLine.addSchedule(newSchedule);
            serviceIdsUsed.push(newServiceId);
            newDuplicatedSchedules[newServiceId] = newSchedule.getAttributes();
        }
    }

    return duplicatedLine;
};
