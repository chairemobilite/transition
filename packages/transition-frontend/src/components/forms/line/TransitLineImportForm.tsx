/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';

import ImporterValidator from 'chaire-lib-common/lib/services/importers/ImporterValidator';
import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';
import FileUploaderHOC, { FileUploaderHOCProps } from 'chaire-lib-frontend/lib/components/input/FileUploaderHOC';
import FileImportForm from '../../parts/FileImportForm';

interface LineImportFormProps extends FileUploaderHOCProps {
    setImporterSelected: (importerSelected: boolean) => void;
}

const LineImportForm: React.FunctionComponent<LineImportFormProps & WithTranslation> = (
    props: LineImportFormProps & WithTranslation
) => {
    const closeImporter = () => props.setImporterSelected(false);

    const onImported = async () => {
        await serviceLocator.collectionManager
            .get('lines')
            .loadFromServer(serviceLocator.socketEventManager, serviceLocator.collectionManager);
        await serviceLocator.collectionManager
            .get('agencies')
            .loadFromServer(serviceLocator.socketEventManager, serviceLocator.collectionManager);
        serviceLocator.collectionManager.refresh('agencies');
        serviceLocator.collectionManager.refresh('lines');
        serviceLocator.eventManager.emit('progress', { name: 'Importing', progress: 1.0 });
        closeImporter();
    };

    React.useEffect(() => {
        props.addEventListeners();
        serviceLocator.socketEventManager.on('importer.linesImported', onImported);
        return () => {
            props.removeEventListeners();
            serviceLocator.socketEventManager.off('importer.linesImported', onImported);
        };
    }, []);

    return (
        <FileImportForm
            validator={props.validator as ImporterValidator}
            pluralizedObjectsName={'lines'}
            fileNameWithExtension={'lines.json'}
            fileUploader={props.fileUploader}
            fileImportRef={props.fileImportRef}
            onChange={props.onChange}
            label={props.t('main:JsonFile')}
            acceptsExtension={'.json'}
            closeImporter={closeImporter}
        />
    );
};

export default FileUploaderHOC(withTranslation(['transit', 'main'])(LineImportForm), ImporterValidator);
