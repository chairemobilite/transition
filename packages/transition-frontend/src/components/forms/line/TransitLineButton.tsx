/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';

import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';
import Line from 'transition-common/lib/services/line/Line';
import Button from '../../parts/Button';
import ButtonCell from '../../parts/ButtonCell';
import { duplicateLine } from 'transition-common/lib/services/line/LineDuplicator';
import { EventManager } from 'chaire-lib-common/lib/services/events/EventManager';
import { MapUpdateLayerEventType } from 'chaire-lib-frontend/lib/services/map/events/MapEventsCallbacks';

interface LineButtonProps extends WithTranslation {
    line: Line;
    selectedLine?: Line;
    lineIsHidden: boolean;
    onObjectSelected?: (objectId: string) => void;
}

const TransitLineButton: React.FunctionComponent<LineButtonProps> = (props: LineButtonProps) => {
    const [lineIsHidden, setLineIsHidden] = React.useState(props.lineIsHidden);
    const lineIsSelected = (props.selectedLine && props.selectedLine.getId() === props.line.getId()) || false;

    const onSelect: React.MouseEventHandler = async (e: React.MouseEvent) => {
        if (e) {
            e.stopPropagation();
        }
        await props.line.refreshSchedules(serviceLocator.socketEventManager);
        props.line.startEditing();
        if (props.onObjectSelected) {
            props.onObjectSelected(props.line.getId());
        }
        serviceLocator.selectedObjectsManager.setSelection('line', [props.line]);
    };

    const onDelete: React.MouseEventHandler = async (e: React.MouseEvent) => {
        if (e) {
            e.stopPropagation();
        }

        const lineHasPaths = props.line.hasPaths();

        serviceLocator.eventManager.emit('progress', { name: 'DeletingLine', progress: 0.0 });
        await props.line.delete(serviceLocator.socketEventManager);
        if (lineIsSelected) {
            serviceLocator.selectedObjectsManager.deselect('line');
            if (lineHasPaths) {
                // reload paths
                await serviceLocator.collectionManager.get('paths').loadFromServer(serviceLocator.socketEventManager);
                serviceLocator.collectionManager.refresh('paths');
                (serviceLocator.eventManager as EventManager).emitEvent<MapUpdateLayerEventType>('map.updateLayer', {
                    layerName: 'transitPaths',
                    data: serviceLocator.collectionManager.get('paths').toGeojson()
                });
            }
        }
        serviceLocator.eventManager.emit('progress', { name: 'DeletingLine', progress: 1.0 });
        serviceLocator.collectionManager.refresh('lines');
    };

    const onDuplicate: React.MouseEventHandler = async (e: React.MouseEvent) => {
        if (e) {
            e.stopPropagation();
        }

        serviceLocator.eventManager.emit('progress', { name: 'SavingLine', progress: 0.0 });
        await duplicateLine(props.line, {
            socket: serviceLocator.socketEventManager,
            duplicateSchedules: true,
            duplicateServices: true,
            newLongname: `${props.line.get('longname')} (${props.t('main:Copy')})`,
            newServiceSuffix: props.t('main:Copy')
        });

        serviceLocator.collectionManager.refresh('paths');
        serviceLocator.collectionManager.refresh('lines');
        serviceLocator.collectionManager.refresh('services');
        (serviceLocator.eventManager as EventManager).emitEvent<MapUpdateLayerEventType>('map.updateLayer', {
            layerName: 'transitPaths',
            data: serviceLocator.collectionManager.get('paths').toGeojson()
        });
        serviceLocator.eventManager.emit('progress', { name: 'SavingLine', progress: 1.0 });
    };

    const showOnMap = (e: React.MouseEvent) => {
        e.stopPropagation();
        if (serviceLocator.keyboardManager.keyIsPressed('alt')) {
            serviceLocator.pathLayerManager.showAllLinesForAgency(props.line.get('agency_id'));
        } else {
            serviceLocator.pathLayerManager.showLineId(props.line.getId());
        }
        setLineIsHidden(false);
    };

    const hideOnMap = (e: React.MouseEvent) => {
        e.stopPropagation();
        if (serviceLocator.keyboardManager.keyIsPressed('alt')) {
            serviceLocator.pathLayerManager.hideAllLinesForAgency(props.line.get('agency_id'));
        } else {
            serviceLocator.pathLayerManager.hideLineId(props.line.getId());
        }
        setLineIsHidden(true);
    };

    const isFrozen = props.line.isFrozen();
    const pathsCount = props.line.paths.length;
    const scheduledServicesCount =
        props.line.attributes.service_ids !== undefined
            ? props.line.attributes.service_ids.length
            : Object.keys(props.line.attributes.scheduleByServiceId).length;

    return (
        <Button
            key={props.line.getId()}
            isSelected={lineIsSelected}
            flushActionButtons={false}
            onSelect={{ handler: onSelect }}
            onDuplicate={{ handler: onDuplicate, altText: props.t('transit:transitLine:DuplicateLine') }}
            onDelete={
                !isFrozen && !lineIsSelected
                    ? {
                        handler: onDelete,
                        message: props.t('transit:transitLine:ConfirmDelete'),
                        altText: props.t('transit:transitLine:Delete')
                    }
                    : undefined
            }
        >
            <ButtonCell alignment="left">
                <span className="_circle-button" style={{ backgroundColor: props.line.attributes.color }}></span>
                {lineIsHidden === true && (
                    <span className="_list-element" onClick={showOnMap} title={props.t('main:Show')}>
                        <img
                            className="_list-element _icon-alone"
                            src={'/dist/images/icons/interface/hidden_white.svg'}
                            alt={props.t('main:Show')}
                            title={props.t('main:Show')}
                        />
                    </span>
                )}
                {lineIsHidden === false && (
                    <span className="_list-element" onClick={hideOnMap} title={props.t('main:Hide')}>
                        <img
                            className="_list-element _icon-alone"
                            src={'/dist/images/icons/interface/visible_white.svg'}
                            alt={props.t('main:Hide')}
                            title={props.t('main:Hide')}
                        />
                    </span>
                )}
                <img
                    className="_list-element _icon-alone"
                    src={`/dist/images/icons/transit/modes/${props.line.getAttributes().mode}_white.svg`}
                    alt={props.t(`transit:transitLine:modes:${props.line.getAttributes().mode}`)}
                    title={props.t(`transit:transitLine:modes:${props.line.getAttributes().mode}`)}
                />
            </ButtonCell>
            {isFrozen && (
                <ButtonCell alignment="left">
                    <img
                        className="_icon-alone"
                        src={'/dist/images/icons/interface/lock_white.svg'}
                        alt={props.t('main:Locked')}
                    />
                </ButtonCell>
            )}
            <ButtonCell alignment="left">{props.line.attributes.shortname}</ButtonCell>
            <ButtonCell alignment="left">{props.line.attributes.longname}</ButtonCell>
            <ButtonCell alignment="flush">
                {pathsCount > 1
                    ? props.t('transit:transitLine:nPaths', { n: pathsCount })
                    : props.t('transit:transitLine:nPath', { n: pathsCount })}{' '}
                {scheduledServicesCount > 0 && (
                    <span className="_list-element">
                        {scheduledServicesCount > 1
                            ? props.t('transit:transitLine:nServices', { n: scheduledServicesCount })
                            : props.t('transit:transitLine:nService', { n: scheduledServicesCount })}
                    </span>
                )}
            </ButtonCell>
        </Button>
    );
};

export default withTranslation(['transit', 'main', 'notifications'])(TransitLineButton);
