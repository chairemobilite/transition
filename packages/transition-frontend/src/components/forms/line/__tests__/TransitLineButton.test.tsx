/*
 * Copyright 2025, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

import * as React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import TransitLineButton from '../TransitLineButton';
import Line from 'transition-common/lib/services/line/Line';
import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';
import * as Status from 'chaire-lib-common/lib/utils/Status';
import { duplicateLine } from 'transition-common/lib/services/line/LineDuplicator';


jest.mock('react-markdown', () => 'Markdown');
jest.mock('remark-gfm', () => 'remark-gfm');

jest.mock('transition-common/lib/services/line/LineDuplicator', () => ({
    duplicateLine: jest.fn(),
}));

jest.mock('chaire-lib-common/lib/utils/ServiceLocator', () => ({
    socketEventManager: {
        emit: jest.fn(),
        on: jest.fn(),
        off: jest.fn(),
    },
    selectedObjectsManager: {
        setSelection: jest.fn(),
        deselect: jest.fn(),
    },
}));

beforeEach(() => {
    jest.clearAllMocks();

    serviceLocator.selectedObjectsManager = {
        setSelection: jest.fn(),
        deselect: jest.fn(),
    };

    serviceLocator.socketEventManager = {
        emit: jest.fn(),
        on: jest.fn(),
        off: jest.fn(),
    };

});

// Wrap tests in a describe block
describe('TransitLineButton', () => {

    test('Minimal props', () => {
        const mockLine = new Line({ some: 'attributes' }, true);
        const { container } = render(
            <TransitLineButton
                line={mockLine}
                lineIsHidden={false}
            >
                test
            </TransitLineButton>
        );

        expect(container).toMatchSnapshot();
    });

    test('All props', () => {
        const mockLine = new Line({ some: 'attributes' }, true);
        const mockOnObjectSelected = (objectId: string) => {};
        const { container } = render(
            <TransitLineButton
                line={mockLine}
                selectedLine={mockLine}
                lineIsHidden={false}
                onObjectSelected={mockOnObjectSelected}
            >
                test
            </TransitLineButton>
        );

        expect(container).toMatchSnapshot();
    });

    test('onSelect should call refreshSchedules, startEditing, and onObjectSelected', async () => {

        const mockLine = new Line({ some: 'attributes' }, true);
        // const mockOnObjectSelected = (objectId: string) => {};
        const mockOnObjectSelected = jest.fn();

        const { getByText } = render(
            <TransitLineButton
                line={mockLine}
                selectedLine={mockLine}
                lineIsHidden={false}
                onObjectSelected={mockOnObjectSelected}
            >
                test
            </TransitLineButton>
        );

        jest.spyOn(mockLine, 'refreshSchedules').mockResolvedValue(undefined);
        jest.spyOn(mockLine, 'getId').mockReturnValue('fake_line_id_1');
        jest.spyOn(mockLine, 'startEditing').mockImplementation(() => {});
        jest.spyOn(mockLine, 'refreshPaths').mockImplementation(() => {});

        const button = document.querySelector('._list-element');
        if (button) fireEvent.click(button);

        await waitFor(() => {
            expect(mockLine.refreshSchedules).toHaveBeenCalledWith(serviceLocator.socketEventManager);
            expect(mockLine.startEditing).toHaveBeenCalled();
            expect(mockOnObjectSelected).toHaveBeenCalledWith('fake_line_id_1');
            expect(serviceLocator.selectedObjectsManager.setSelection).toHaveBeenCalledWith('line', [mockLine]);
        });
    });

    test('onDelete should trigger delete when the line is NOT selected', async () => {

        // First mocked line (the one being tested for deletion)
        const mockLine = new Line({ some: 'attributes' }, true);
        jest.spyOn(mockLine, 'hasPaths').mockReturnValue(true);
        jest.spyOn(mockLine, 'isFrozen').mockReturnValue(false);
        jest.spyOn(mockLine, 'delete').mockResolvedValue(
            Status.createOk({ id: 'fake_line_id_1' })
        );
        jest.spyOn(mockLine, 'getId').mockReturnValue('fake_line_id_1');

        // Second mocked line (the currently selected line)
        const mockLine2 = new Line({ some: 'other_attributes' }, true);
        jest.spyOn(mockLine2, 'getId').mockReturnValue('fake_line_id_2');

        // Mock serviceLocator
        serviceLocator.selectedObjectsManager = {
            setSelection: jest.fn(),
            deselect: jest.fn(),
        };
        serviceLocator.collectionManager = {
            get: jest.fn().mockImplementation((name) => {
                if (name === 'paths') {
                    return {
                        loadFromServer: jest.fn().mockResolvedValue(undefined),
                        toGeojson: jest.fn().mockReturnValue({}),
                    };
                }
                return { refresh: jest.fn() };
            }),
            refresh: jest.fn(),
        };
        serviceLocator.eventManager = {
            emit: jest.fn(),
            emitEvent: jest.fn(),
        };
        serviceLocator.socketEventManager = {
            emit: jest.fn(),
            off: jest.fn(),
            on: jest.fn(),
        };


        const { container, getByText } = render(
            <TransitLineButton
                line={mockLine}
                selectedLine={mockLine2}
                lineIsHidden={false}
                onObjectSelected={jest.fn()}
            />
        );

        // Verify that the delete button exists
        const deleteButton = container.querySelector('img[alt="transit:transitLine:Delete"]');
        expect(deleteButton).not.toBeNull();
        fireEvent.click(deleteButton as Element);

        // Confirm delete in the modal
        const confirmDeleteButton = await waitFor(() =>
            getByText('transit:transitLine:Delete')
        );
        fireEvent.click(confirmDeleteButton);

        // Ensure delete process executes correctly
        await waitFor(() => {
            expect(mockLine.delete).toHaveBeenCalledWith(serviceLocator.socketEventManager);
            expect(serviceLocator.collectionManager.refresh).toHaveBeenCalledWith('lines');
        });
    });

    test('onDuplicate should call duplicateLine and refresh collections', async () => {
        // Mocking a line
        const mockLine = new Line({ some: 'attributes', longname: 'Test Line' }, true);
        jest.spyOn(mockLine, 'get').mockImplementation((key) => {
            if (key === 'longname') return 'Test Line';
            if (key === 'is_frozen') return false; // Mock is_frozen attribute
            return undefined;
        });

        // Mock isFrozen method
        mockLine.isFrozen = jest.fn().mockReturnValue(false); // Default to not frozen

        // Mock getAgency method if it exists
        mockLine.getAgency = jest.fn().mockReturnValue({
            get: jest.fn().mockImplementation((key) => {
                if (key === 'is_frozen') return false; // Mock agency is_frozen attribute
                return undefined;
            }),
        });

        // Mocking duplicateLine to return a new line instance
        const duplicatedLine = new Line({ some: 'newAttributes', longname: 'Test Line (Copy)' }, true);
        (duplicateLine as jest.Mock).mockResolvedValue(duplicatedLine);

        // Mock ServiceLocator functions
        serviceLocator.collectionManager = {
            get: jest.fn((collectionName) => {
                if (collectionName === 'agencies') {
                    return {
                        getById: jest.fn().mockReturnValue({ id: 'fake_agency_id' }), // Mock agency
                    };
                }
                if (collectionName === 'paths') {
                    return {
                        toGeojson: jest.fn().mockReturnValue({}),
                    };
                }
                return { refresh: jest.fn() }; // Default return
            }),
            refresh: jest.fn(),
        };

        serviceLocator.eventManager = {
            emit: jest.fn(),
            emitEvent: jest.fn(),
        };
        serviceLocator.socketEventManager = {};

        // Render the component with hideActions set to false so the duplicate button is visible
        const { container } = render(
            <TransitLineButton
                line={mockLine}
                selectedLine={mockLine}
                lineIsHidden={false}
                onObjectSelected={jest.fn()}
            />
        );

        // Find the duplicate button using querySelector (Adjust selector as needed)
        const duplicateButton = container.querySelector('img[alt="transit:transitLine:DuplicateLine"]');
        expect(duplicateButton).toBeInTheDocument(); // Ensure the button is found
        if(duplicateButton) {
            fireEvent.click(duplicateButton); // Click the button
        }

    });
});
