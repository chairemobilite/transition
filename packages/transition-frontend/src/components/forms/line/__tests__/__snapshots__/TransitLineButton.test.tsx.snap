// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitLineButton All props 1`] = `
<div>
  <li
    class="_list _active"
  >
    <span
      class="_list-group _left"
    >
      <span
        class="_list-element"
      >
        <span
          class="_circle-button"
        />
        <span
          class="_list-element"
          title="main:Hide"
        >
          <img
            alt="main:Hide"
            class="_list-element _icon-alone"
            src="/dist/images/icons/interface/visible_white.svg"
            title="main:Hide"
          />
        </span>
        <img
          alt="transit:transitLine:modes:null"
          class="_list-element _icon-alone"
          src="/dist/images/icons/transit/modes/null_white.svg"
          title="transit:transitLine:modes:null"
        />
      </span>
    </span>
    <span
      class="_list-group _left"
    >
      <span
        class="_list-element"
      />
    </span>
    <span
      class="_list-group _left"
    >
      <span
        class="_list-element"
      />
    </span>
    <span
      class="_list-group _flush-right _right"
    >
      <span
        class="_list-element"
      >
        transit:transitLine:nPath
         
      </span>
    </span>
    <span
      class="_list-group _right"
      title="main:Edit"
    >
      <span
        class="_list-element"
      >
        <img
          alt="main:Edit"
          class="_icon-alone"
          src="/dist/images/icons/interface/edit_white.svg"
        />
      </span>
    </span>
    <span
      class="_list-group _right"
      title="transit:transitLine:DuplicateLine"
    >
      <span
        class="_list-element"
      >
        <img
          alt="transit:transitLine:DuplicateLine"
          class="_icon-alone"
          src="/dist/images/icons/interface/copy_white.svg"
        />
      </span>
    </span>
  </li>
  <li
    class="_clear"
  />
</div>
`;

exports[`TransitLineButton Minimal props 1`] = `
<div>
  <li
    class="_list"
  >
    <span
      class="_list-group _left"
    >
      <span
        class="_list-element"
      >
        <span
          class="_circle-button"
        />
        <span
          class="_list-element"
          title="main:Hide"
        >
          <img
            alt="main:Hide"
            class="_list-element _icon-alone"
            src="/dist/images/icons/interface/visible_white.svg"
            title="main:Hide"
          />
        </span>
        <img
          alt="transit:transitLine:modes:null"
          class="_list-element _icon-alone"
          src="/dist/images/icons/transit/modes/null_white.svg"
          title="transit:transitLine:modes:null"
        />
      </span>
    </span>
    <span
      class="_list-group _left"
    >
      <span
        class="_list-element"
      />
    </span>
    <span
      class="_list-group _left"
    >
      <span
        class="_list-element"
      />
    </span>
    <span
      class="_list-group _flush-right _right"
    >
      <span
        class="_list-element"
      >
        transit:transitLine:nPath
         
      </span>
    </span>
    <span
      class="_list-group _right"
      title="main:Edit"
    >
      <span
        class="_list-element"
      >
        <img
          alt="main:Edit"
          class="_icon-alone"
          src="/dist/images/icons/interface/edit_white.svg"
        />
      </span>
    </span>
    <span
      class="_list-group _right"
      title="transit:transitLine:DuplicateLine"
    >
      <span
        class="_list-element"
      >
        <img
          alt="transit:transitLine:DuplicateLine"
          class="_icon-alone"
          src="/dist/images/icons/interface/copy_white.svg"
        />
      </span>
    </span>
    <span
      class="_list-group _right"
      title="transit:transitLine:Delete"
    >
      <span
        class="_list-element"
      >
        <img
          alt="transit:transitLine:Delete"
          class="_icon-alone"
          src="/dist/images/icons/interface/delete_white.svg"
        />
      </span>
    </span>
  </li>
  <li
    class="_clear"
  />
</div>
`;
